File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
public int numIslands(char[][] grid) {
        int islands = 0;
        // use 2 for loops to iterate thru the grid, if we find '1' then we search for the adjacent land
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == '1') {
                    islands++;
                }
                islandSearch(grid, i, j);
            }
        }
        return islands;

    }

    public void islandSearch(char[][] grid, int i, int j) {
        // recursively search the size of an island
        // base conditions: outside boundary or is not a 1
        if (i < 0 || i > grid.length - 1 || j < 0 || j > grid[i].length - 1 || grid[i][j] == '0') {
            return;
        } else {
            // we set already searched index to 0, then recursively search the adjacent index for the whole island
            grid[i][j] = '0';
            islandSearch(grid, i + 1, j);
            islandSearch(grid, i, j + 1);
            islandSearch(grid, i - 1, j);
            islandSearch(grid, i, j - 1);
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        #helper function to dfs and set island cells to '0'
        def dfs(i,j):
            #check out of bounds, or not '1'
            if i<0 or j<0 or i>=len(grid) or j>=len(grid[0]) or grid[i][j] != '1':
                return
            grid[i][j] = '0'
            dfs(i,j+1)
            dfs(i,j-1)
            dfs(i-1,j)
            dfs(i+1,j)
        islands = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    islands+=1
                    dfs(i,j)
        return islands